File: gaussian_basis/gaussian_basis/gaussian3d.py
from .gaussian1d import Gaussian1D
import numpy as np


class Gaussian3D:
    pos: np.ndarray
    ang: np.ndarray
    orb_exp: float
    amp: float
    gaussian0: Gaussian1D
    gaussian1: Gaussian1D
    gaussian2: Gaussian1D

    def __init__(self, amp: float, orb_exp: float,
                 pos: np.ndarray, ang: np.ndarray):
        if pos.shape != (3,) or ang.shape != (3,):
            raise Exception
        self.amp = amp
        self.pos = pos
        self.ang = ang
        self.orb_exp = orb_exp
        self.gaussian0 = Gaussian1D(orb_exp, pos[0], int(ang[0]))
        self.gaussian1 = Gaussian1D(orb_exp, pos[1], int(ang[1]))
        self.gaussian2 = Gaussian1D(orb_exp, pos[2], int(ang[2]))

    def __call__(self, r: np.ndarray) -> np.ndarray:
        return self.amp*self.gaussian0(r[0]) * \
               self.gaussian1(r[1])*self.gaussian2(r[2])

    def __mul__(self, other: float) -> 'Gaussian3D':
        return Gaussian3D(other*self.amp,
                          self.orb_exp, self.pos, self.ang)

    def __rmul__(self, other: float) -> 'Gaussian3D':
        return Gaussian3D(other*self.amp,
                          self.orb_exp, self.pos, self.ang)

    def get_gaussian1d(self, index: int) -> Gaussian1D:
        if index == 0:
            return self.gaussian0
        elif index == 1:
            return self.gaussian1
        elif index == 2:
            return self.gaussian2
        else:
            raise IndexError

    def position(self):
        return self.pos

    def orbital_exponent(self):
        return self.orb_exp

    def angular(self):
        return self.ang

    def amplitude(self):
        return self.amp


def product_center(g: Gaussian3D, h: Gaussian3D) -> np.ndarray:
    return (g.orbital_exponent()*g.position()
            + h.orbital_exponent()*h.position()) / \
           (g.orbital_exponent() + h.orbital_exponent())

File: gaussian_basis/gaussian_basis/src/gaussian3d.cpp
#include <cmath>
#include "gaussian3d.hpp"


Gaussian1D Gaussian3D::get_gaussian1d(int index) const {
    if (index == 0) {
        return Gaussian1D(orb_exp, this->r0[0], (long)this->ang[0]);
        // return gaussian0;
    } else if (index == 1) {
        return Gaussian1D(orb_exp, this->r0[1], (long)this->ang[1]);
        // return gaussian1;
    } else if (index == 2) {
        return Gaussian1D(orb_exp, this->r0[2], (long)this->ang[2]);
        // return gaussian2;
    } else {
        return {};
    }
}

Vec3 Gaussian3D::position() const {
    return this->r0;
    // return {{{this->gaussian0.position(),
    //         this->gaussian1.position(),
    //         this->gaussian2.position()}}};
}

double Gaussian3D::orbital_exponent() const {
    return this->orb_exp;
}

Vec3 Gaussian3D::angular() const {
    return {{{
        (double)this->ang[0], (double)this->ang[1], (double)this->ang[2]}}};
    // return {{{(double)this->gaussian0.angular(),
    //         (double)this->gaussian1.angular(),
    //         (double)this->gaussian2.angular()}}};
}

double Gaussian3D::amplitude() const {
    return this->amp;
}

struct Vec3 product_center(const Gaussian3D &g, const Gaussian3D &h) {
    return ((g.orbital_exponent()*g.position())
            + (h.orbital_exponent()*h.position())) / 
           (g.orbital_exponent() + h.orbital_exponent());
}
